lib/twitter_cldr/utils/script_detector.rb

```diff
@@ -0,0 +1,81 @@
+# encoding: UTF-8
+
+# Copyright 2012 Twitter, Inc
+# http://www.apache.org/licenses/LICENSE-2.0
+
+module TwitterCldr
+  module Utils
+    class ScriptDetectionResult
+
+      attr_reader :scores
+
+      def initialize(scores)
+        @scores = scores
+      end
+
+      def best_guess
+        max_score = 0
+        max_script_name = nil
+
+        scores.each_pair do |script_name, score|
+          if score > max_score
+            max_score = score
+            max_script_name = script_name
+          end
+        end
+
+        max_script_name
+      end
+
+      def score_for(script_name)
+        scores[script_name]
+      end
+
+    end
+
+    class ScriptDetector
+
+      class << self
+
+        def detect_scripts(text)
+          length = text.length.to_f
+
+          ScriptDetectionResult.new(
+            scores_for(text).each_with_object({}) do |(script_name, count), ret|
+              ret[script_name] = count / length
+            end
+          )
+        end
+
+        protected
```

Comment:
Well, Ruby's `protected` is quite different from Java's. It has nothing to do with inheritance and its only difference from `private` is that you can call protected methods with explicit receiver if the caller is of the same type. Basically it allows an object to call its own methods as `self.my_method` (which is bad style anyway) and it allows an object to call protected methods of other objects as long as they are instances of the same class (rarely needed). 

Here's a [more detailed explanation](http://nithinbekal.com/posts/ruby-protected-methods/) (with the only use case for it I've ever seen).